api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -227,132 +360,200 @@ private void maybeAlertOnCostThresholds(
+   * so we need to recalculate it 2.4. Free Tier Usage time is after the Workspace last updated
+   * time, but the difference is smaller than a certain value. This case to account for charges that
+   * may occur after the workspace gets deleted and after the last cron had run.
    *
-   * @param user the user as represented in our database
-   * @return the total USD amount spent in workspaces created by this user, represented as a double
+   * @param allCostsInDbForUsers
+   * @return a {@link java.util.List} of the workspaces that require updates only
    */
-  @Nullable
-  public Double getCachedFreeTierUsage(DbUser user) {
-    return workspaceFreeTierUsageDao.totalCostByUser(user);
+  @NotNull
+  private List<WorkspaceCostView> findActiveOrDeletedWorkspacesThatRequireUpdates(
+      List<WorkspaceCostView> allCostsInDbForUsers) {
+
+    List<WorkspaceCostView> workspacesThatRequireUpdate =
+        allCostsInDbForUsers.stream()
+            .filter(
+                c ->
+                    (c.getActiveStatus() == 0)
+                        || (c.getActiveStatus() == 1
+                            && (c.getFreeTierLastUpdated() == null
+                                || c.getWorkspaceLastUpdated() == null
+                                || c.getFreeTierLastUpdated().before(c.getWorkspaceLastUpdated())
+                                || (c.getFreeTierLastUpdated().after(c.getWorkspaceLastUpdated())
+                                    && c.getFreeTierLastUpdated().getTime()
+                                            - c.getWorkspaceLastUpdated().getTime()
+                                        < getNumberOfDaysToConsiderForFreeTierUsageUpdate()))))
+            .collect(Collectors.toList());
+
+    logger.info(
+        String.format("Workspaces that require update %d", workspacesThatRequireUpdate.size()));
+
+    return workspacesThatRequireUpdate;
   }
 
   /**
-   * Does this user have remaining free tier credits? Compare the user-specific free tier limit (may
-   * be the system default) to the amount they have used.
+   * Requesting live costs from BQ in batches then updating it in a transactional method.
    *
-   * @param user the user as represented in our database
-   * @return whether the user has remaining credits
+   * @param allCostsInDbForUsers
+   * @param workspacesThatRequireUpdate
+   * @return a Map of all live costs.
    */
-  public boolean userHasRemainingFreeTierCredits(DbUser user) {
-    final double usage = Optional.ofNullable(getCachedFreeTierUsage(user)).orElse(0.0);
-    return !costAboveLimit(user, usage);
+  private Map<Long, Double> updateFreeTierUsageInBatches(
+      List<WorkspaceCostView> allCostsInDbForUsers,
+      List<WorkspaceCostView> workspacesThatRequireUpdate) {
+    final Map<Long, Double> liveCostsInBQ = new HashMap<>();
+    for (List<WorkspaceCostView> workspacesPartition :
+        Lists.partition(workspacesThatRequireUpdate, 1000)) {
```

Comment:
what is the 1000? Should it be static constant?

Reply:
I am thinking of making this a config parameter. In prod, this won't happen a lot I believe, but we have several users in test with thousands of workspaces.
1000 because during my testing, BQ queries took very long for anything greater than that, causing timeouts. So it's mainly based on my testing. 
Do you think it should be config parameter?

Reply:
config is useful to customize per env and for fast iteration/upgrade. 
There is a balance between making everything config them managing a giant config file vs having minimum configs. 
To me if that is a value that barely change, not necessary to have different value per env, then it can just be a constant.